lte/gateway/c/core/oai/tasks/amf/amf_config.c

```diff
@@ -0,0 +1,584 @@
+            calloc(num, sizeof(*config_pP->served_tai.plmn_mnc_len));
+        config_pP->served_tai.tac =
+            calloc(num, sizeof(*config_pP->served_tai.tac));
+      }
+
+      config_pP->served_tai.nb_tai = num;
+
+      for (i = 0; i < num; i++) {
+        sub2setting = config_setting_get_elem(setting, i);
+
+        if (sub2setting != NULL) {
+          if ((config_setting_lookup_string(
+                  sub2setting, MME_CONFIG_STRING_MCC, &mcc))) {
+            config_pP->served_tai.plmn_mcc[i] = (uint16_t) atoi(mcc);
+          }
+
+          if ((config_setting_lookup_string(
+                  sub2setting, MME_CONFIG_STRING_MNC, &mnc))) {
+            config_pP->served_tai.plmn_mnc[i]     = (uint16_t) atoi(mnc);
+            config_pP->served_tai.plmn_mnc_len[i] = strlen(mnc);
+
+            AssertFatal(
+                (config_pP->served_tai.plmn_mnc_len[i] == MIN_MNC_LENGTH) ||
+                    (config_pP->served_tai.plmn_mnc_len[i] == MAX_MNC_LENGTH),
+                "Bad MNC length %u, must be %d or %d",
+                config_pP->served_tai.plmn_mnc_len[i], MIN_MNC_LENGTH,
+                MAX_MNC_LENGTH);
+          }
+
+          if ((config_setting_lookup_string(
+                  sub2setting, MME_CONFIG_STRING_TAC, &tac))) {
+            config_pP->served_tai.tac[i] = (uint16_t) atoi(tac);
+
+            if (!TAC_IS_VALID(config_pP->served_tai.tac[i])) {
+              fprintf(
+                  stderr, "ERROR: Invalid TAC value " TAC_FMT,
+                  config_pP->served_tai.tac[i]);
+            }
+          }
+        }
+      }
+
+      // sort TAI list
+      n = config_pP->served_tai.nb_tai;
+      do {
+        stop_index = 0;
+        for (i = 1; i < n; i++) {
+          swap = false;
+          if (config_pP->served_tai.plmn_mcc[i - 1] >
+              config_pP->served_tai.plmn_mcc[i]) {
+            swap = true;
+          } else if (
+              config_pP->served_tai.plmn_mcc[i - 1] ==
+              config_pP->served_tai.plmn_mcc[i]) {
+            if (config_pP->served_tai.plmn_mnc[i - 1] >
+                config_pP->served_tai.plmn_mnc[i]) {
+              swap = true;
+            } else if (
+                config_pP->served_tai.plmn_mnc[i - 1] ==
+                config_pP->served_tai.plmn_mnc[i]) {
+              if (config_pP->served_tai.tac[i - 1] >
+                  config_pP->served_tai.tac[i]) {
+                swap = true;
+              }
+            }
+          }
+          if (true == swap) {
+            uint16_t swap16;
+            swap16 = config_pP->served_tai.plmn_mcc[i - 1];
+            config_pP->served_tai.plmn_mcc[i - 1] =
+                config_pP->served_tai.plmn_mcc[i];
+            config_pP->served_tai.plmn_mcc[i] = swap16;
+
+            swap16 = config_pP->served_tai.plmn_mnc[i - 1];
+            config_pP->served_tai.plmn_mnc[i - 1] =
+                config_pP->served_tai.plmn_mnc[i];
+            config_pP->served_tai.plmn_mnc[i] = swap16;
+
```

Comment:
you need to also do:
```
            swap16 = config_pP->served_tai.plmn_mnc_len[i - 1];
            config_pP->served_tai.plmn_mnc_len[i - 1] =
                config_pP->served_tai.plmn_mnc_len[i];
            config_pP->served_tai.plmn_mnc_len[i] = swap16;
```

check out the most up to date `mme_config.c`. I previously commented that brute force copying the implementation from mme side would carry over the bugs and fixes on the MME side will not automatically carry over here. You need to refactor amf configuration such that parsing happens only once. I suggest to remove all the common configs that are already parsed by mme config and define another function that you would call during amf config initialization that simply copies them over rather than parsing again.

Reply:
Separate Zenhub ticket(#8332) opened to handle

Reply:
hey could you fix the bug here until the other zenhub issue is resolved?

Reply:
resolved and changes commited